Python/marshal.c

```diff
@@ -895,51 +988,40 @@ r_PyLong(RFILE *p)
         return NULL;
     }
 
-    size = 1 + (Py_ABS(n) - 1) / PyLong_MARSHAL_RATIO;
-    shorts_in_top_digit = 1 + (Py_ABS(n) - 1) % PyLong_MARSHAL_RATIO;
-    ob = _PyLong_New(size);
-    if (ob == NULL)
-        return NULL;
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+    Py_ssize_t marshal_ratio = layout->bits_per_digit/PyLong_MARSHAL_SHIFT;
 
-    _PyLong_SetSignAndDigitCount(ob, n < 0 ? -1 : 1, size);
+    /* must be a multiple of PyLong_MARSHAL_SHIFT */
+    assert(layout->bits_per_digit % PyLong_MARSHAL_SHIFT == 0);
 
-    for (i = 0; i < size-1; i++) {
-        d = 0;
-        for (j=0; j < PyLong_MARSHAL_RATIO; j++) {
-            md = r_short(p);
-            if (md < 0 || md > PyLong_MARSHAL_BASE)
-                goto bad_digit;
-            d += (digit)md << j*PyLong_MARSHAL_SHIFT;
-        }
-        ob->long_value.ob_digit[i] = d;
+    /* other assumptions on PyLongObject internals */
+    assert(layout->bits_per_digit <= 32);
+    assert(layout->digits_order == -1);
+    assert(layout->digit_endianness == (PY_LITTLE_ENDIAN ? -1 : 1));
+    assert(layout->digit_size == 2 || layout->digit_size == 4);
+
+    Py_ssize_t size = 1 + (Py_ABS(n) - 1) / marshal_ratio;
```

Comment:
You need a special case for n==0, like the code that you removed, no?

Reply:
If n==0, `1 + (Py_ABS(n) - 1) / marshal_ratio` gives `size=0`, no?

`size=0` is annoying. `_w_digitsXX()` functions use `size - 1` for example.

Please keep `if (n == 0) return (PyObject *)_PyLong_New(0);` to avoid this annoying case.

> But 0 has one digit allocated anyway. So, all works correctly:

Ok, but I'm thinking about invalid/special marshal data. I would prefer to avoid a crash if possible.

Reply:
> if n==0, 1 + (Py_ABS(n) - 1) / marshal_ratio gives size=0, no?

No.  (-1)/positive is 0 in C.

> Please keep if (n == 0) return (PyObject *)_PyLong_New(0)

Then we will depend on this import.

> I'm thinking about invalid/special marshal data.

An example?

Reply:
> No. (-1)/positive is 0 in C.

Oh, I forgot that. It might help future readers to add a comment explaining that size=1 if n=0.

Reply:
Maybe add also an assertion: `assert(size >= 1)`.